File: 07/taehyuk/690.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        employeeDic = {employee.id: employee for employee in employees}

        def DFS(id):
            importance = employeeDic[id].importance            
            for subordinateId in employeeDic[id].subordinates:
                importance += DFS(subordinateId)
            return importance
        
        return DFS(id)
        

File: 07/JYC/690.ts
/**
 * Definition for Employee.
 * class Employee {
 *     id: number
 *     importance: number
 *     subordinates: number[]
 *     constructor(id: number, importance: number, subordinates: number[]) {
 *         this.id = (id === undefined) ? 0 : id;
 *         this.importance = (importance === undefined) ? 0 : importance;
 *         this.subordinates = (subordinates === undefined) ? [] : subordinates;
 *     }
 * }
 */

function getImportance(employees: Employee[], id: number): number {
  const employeeMap = new Map<number, Employee>();
  for (const employee of employees) {
    employeeMap.set(employee.id, employee);
  }

  function dfs(empId: number): number {
    const employee = employeeMap.get(empId);
    if (!employee) return 0;

    let totalImportance = employee.importance;

    for (const subId of employee.subordinates) {
      totalImportance += dfs(subId);
    }

    return totalImportance;
  }

  return dfs(id);
}
